libfwupdplugin/fu-common.c

```diff
@@ -3384,3 +3384,90 @@ fu_xmlb_builder_insert_kb (XbBuilderNode *bn, const gchar *key, gboolean value)
 {
 	xb_builder_node_insert_text (bn, key, value ? "true" : "false", NULL);
 }
+
+/**
+ * fu_common_read_firmware_search_path:
+ * @error: (nullable): optional return location for an error
+ *
+ * Reads the FU_PATH_KIND_FIRMWARE_SEARCH and
+ * returns it's contents
+ *
+ * Returns: a pointer to a gchar array of PATH_MAX
+ *
+ * Since: 1.6.2
+ **/
+gchar *
+fu_common_read_firmware_search_path (GError **error)
+{
+	FILE *fd;
+	gsize res = 0;
+	g_autofree gchar *sys_fw_search_path_prm = NULL;
+	g_autofree gchar *path = g_malloc0 (PATH_MAX);
+
+	/* write mm_fw_dir to a search path parameter of the firmware_loader module */
+	sys_fw_search_path_prm = fu_common_get_path (FU_PATH_KIND_FIRMWARE_SEARCH);
+	/* g_file_set_contents will try to create backup files in sysfs, so use fopen here */
+	fd = fopen (sys_fw_search_path_prm, "r");
+	if (fd == NULL) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_PERMISSION_DENIED,
+			     "Failed to open %s: %s",
+			     sys_fw_search_path_prm,
+			     g_strerror (errno));
+		return NULL;
+	}
+	res = fread (path, sizeof (gchar), PATH_MAX, fd);
+
+	g_debug ("%s(): read %lu bytes: %s", __func__, res, path);
+
+	fclose (fd);
+
+	return g_steal_pointer (&path);
+}
+
+/**
+ * fu_common_write_firmware_search_path:
+ * @path: string to write
+ * @error: (nullable): optional return location for an error
+ *
+ * Writes to the FU_PATH_KIND_FIRMWARE_SEARCH
+ *
+ * Returns: %TRUE if successful
+ *
+ * Since: 1.6.2
+ **/
+gboolean
+fu_common_write_firmware_search_path (const gchar *path, GError **error)
+{
+	FILE *fd;
+	gsize res = 0;
+	g_autofree gchar *sys_fw_search_path_prm = NULL;
+
+	/* write mm_fw_dir to a search path parameter of the firmware_loader module */
+	sys_fw_search_path_prm = fu_common_get_path (FU_PATH_KIND_FIRMWARE_SEARCH);
+	/* g_file_set_contents will try to create backup files in sysfs, so use fopen here */
+	fd = fopen (sys_fw_search_path_prm, "w");
+	if (fd == NULL) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_PERMISSION_DENIED,
+			     "Failed to open %s: %s",
+			     sys_fw_search_path_prm,
+			     g_strerror (errno));
+		return FALSE;
+	}
+	res = fwrite (path, sizeof (gchar), strlen (path), fd);
```

Comment:
g_file_set_contents() maybe, instead of all this custom code?

Reply:
`g_file_set_contents()` will try to create backup files during writing, which is not allowed for sysfs. I think it should be possible with `g_file_set_contents_full` and `G_FILE_SET_CONTENTS_NONE`, but its only available in the latest version of GLib.

Reply:
Maybe there's a different preferred way to do this in fwupd?

Reply:
i could not find anything similar in the codebase. maybe @hughsie and @superm1 can suggest something

Reply:
My opinion, how about check the glib version with a macro and use `g_file_set_contents_full` if 2.66 or later.  And then use this code if older version.  Eventually some day we bump up minimum glib version requirement and can drop the dead code.

Reply:
okay, I have added `g_file_set_contents_full` with `GLIB_CHECK_VERSION(2,66,0)`